### DISCUSSION TITLE
How can I get coordinates of nodes/junction?

### DISCUSSION THREAD

**Discussion by nnet55** [pyswmm/pyswmm#433] [General]
How can I get the coordinates of nodes/junctions in python and store them in a matrix with node number and their coordinates?

**Comment by bemcdonnell** [comment]
@nnet55 I suggest you parse the INP file, directly or use `swmmio` to fetch those values out.  They are not technically related to running a simulation so those values haven't been formally exposed through PySWMM.  However, if would be interested in adding that feature, we can talk about options!

**Reply by nnet55** [reply]
My main intention of using pyswmm was to link  GIS features in SWMM, for coordinating between two platforms I thought that spatial information would be a great addition. I am a beginner in  this so if you share more information, it will be very useful. 
As of now I am planning to use "hymo" for the same.

**Reply by bemcdonnell** [reply]
@nnet55 There are some larger plans underway of migrating the INP text file to use the geopackage standard - sqlite3.   I am open to exposing the XY coords for Nodes here in PySWMM but I don't think introducing any GIS capabilities would be that useful (e.g. functions and coord reference systems).

**Comment by dickinsonre** [comment]
Historically, the engine of SWMM has never use the coordinates so they are not normally exported - they are only used in GUI's.  Do you have an engine idea that would use the coordinates in some manner?

**Reply by nnet55** [reply]
It can give a huge platform for GIS integration.

**Comment by swmm-js** [comment]
If the pyswmm engine utilized doppler or other imaging rainfall, then catchment-specific results can be created for real-time storms. I think I spoke about this with Luis.

swmmReact/swmmNode modifies the .inp file to store geospatial coordinates without breaking the parsing engine, and pyswmm could do something similar. There's a thousand other uses for geospatial integration, it is really important.

**Comment by swmm-js** [comment]
Here's an example of how you can store a geospatial transform in an .inp file. I'll write up something a little more in-depth and show you how you can integrate something like this in pyswmm.

;[PCS]
;;Name           proj      zone                datum              units              defs   
;;------------------------------------------------------------------------------------------------------------------
swmmNode_proj    utm        44                 WGS84              ft                 no_defs    


;[TRANSFORM]
;;Name           x                    y                    size                 rotation  
;;------------------------------------------------------------------------------------------------------------------
swmmNode_t       1283304.9924983545   5681005.354355253    1                    -3.14159

**Comment by bemcdonnell** [comment]
@swmm-js I think that use case is interesting but it fits more as a plug-in rather than a required pyswmm feature to on-board.  The engine itself will not 'necessarily' benefit from having a data model to support coordinates, directly, since it is not used in any calculation.  Runoff surfaces and Nodes have centroids / coordinates which can be mapped geospatially.  SWMM/PySWMM projects really do not need to care about the geospatial aspects until perhaps we migrate the INP to use geopackage standard.  A plugin could be used to map the Subcatchment location to rainfall series cleaner than onboarding these as native features.

**Reply by Jannik-Schilling** [reply]
> (...) until perhaps we migrate the INP to use geopackage standard.

That would be fantastix. I´ve been writing a QGIS [plugin](https://github.com/Jannik-Schilling/generate_swmm_inp) to get the data into QGIS (where you have tons of useful tools to change your data based on spatial information) and back into SWMM. However, it is in some parts still a workaround. And one major drawback is the need to know the coordinate reference system (-> EPSG Code) or at least guess a suitable coordinate reference system in order to display the SWMM features in layers.

**Comment by swmm-js** [comment]
I have built a ton of geospatial databases for swmm using sqlite and spatialLite and created the swmm translator used by a few municipalities, back in 2012 maybe? Probably earlier. It is a great database system, a good choice, and I can't wait to see PySWMM implement it.

**Reply by bemcdonnell** [reply]
@swmm-js Could you open up a discussion in the OWA-SWMM repo to talk on this topic? I think your support would be amazing.  SWMM is heading in a really great direction and all we need is help to make it all happen!

**Comment by swmm-js** [comment]
Absolutely. I'm working on some interactive animated geospatial maps right now, but when I'm done I'll try to write something concise and meaningful.
